src/ucs/vfs/sock/vfs_sock.c

```diff
@@ -170,3 +188,185 @@ int ucs_vfs_sock_recv(int sockfd, ucs_vfs_sock_message_t *vfs_msg)
 
     return 0;
 }
+
+int ucs_vfs_data_destroy_shared(ucs_vfs_data_t *data)
+{
+    int destroyed = 0;
+    struct shmid_ds stat;
+    int ret;
+
+    ucs_debug("destroy VFS shared data %p", data);
+    if (data->shared != UCS_VFS_UNDEFINED) {
+        ret = shmdt(data->shared);
+        if (ret != 0) {
+            ucs_warn("shmdt(%p) failed: %m", data->shared);
+        }
+    }
+
+    if (data->shmid != -1) {
+        /* Check how many existing attachments exist from other processes */
+        ret = shmctl(data->shmid, IPC_STAT, &stat);
+        if (ret != 0) {
+            ucs_warn("shmctl(%d, IPC_STAT) failed: %m", data->shmid);
+        } else if (stat.shm_nattch == 0) {
```

Comment:
but if a process is killed before getting to this part?

Reply:
i am not sure it's ok, some systems may be sensitive to resources remaining in use after exit. especially that this may happen when ucx_vfs daemon was not used at all

Reply:
Well, for the process kill scenario we may setup a signal handler.
Only in case of a crash we will have a static resource leak, but this we cannot workaround I'm afraid with current approach.

Reply:
We need an approach that will not result in a resource leak in case of a crash. At least when VFS was just enabled and not actually used.

Reply:
Yes, on that cluster (NVHPC ice4) `/dev/shm` is not that full:
```
iyastrebov@ice4:~> ll /tmp | wc -l
39962
iyastrebov@ice4:~> ./test /tmp
inotify_add_watch took 1070417 us
iyastrebov@ice4:~> ll /dev/shm | wc -l
386
iyastrebov@ice4:~> ./test /dev/shm
inotify_add_watch took 65 us
```
So this solution would work on that particular cluster.
But IMO this is not a real solution to the problem, we just move the problem to another place.
1. We cannot guarantee that this folder is that free on other machines. Btw does it always exist on all platforms?
2. We still suffer from unrelated events in that folder

inotify on "public" folder has proved to cause an unpredictable delay, which is hard to detect (unless it's that visible, like in this case in NVHPC cluster). This problem already happened, so I think we need a proper solution:
1. inotify on `/tmp/ucx-vfs-$user/` folder. This way we never intersect with the rest of the files in /tmp
Yes, I remember that you don't want client to create folders, but it's always a tradeoff..
2. Use shared memory approach (or maybe pipe)

Let's discuss and decide on this